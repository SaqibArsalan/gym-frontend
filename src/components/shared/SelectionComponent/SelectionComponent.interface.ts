export interface ISelectionComponentProps {
	id?: string;
	keyValue?: string;
	label?: string;
	placeholder?: string;
	defaultValue?: string;
	values?: any[];
	disabled?: boolean;
	itemDisplayKey?: string;
	errorText?: string;
	onChange?: (value: { [key: string]: string } | string) => void;
	overrideInputStyles?: any;
	overrideSelectStyles?: any;
}
