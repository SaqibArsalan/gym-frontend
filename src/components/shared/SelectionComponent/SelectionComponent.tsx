import React, { useState } from 'react';

import {
	FormControl,
	FormHelperText,
	FormLabel,
	MenuItem,
	OutlinedInput,
	Select,
} from '@mui/material';
import { ISelectionComponentProps } from './SelectionComponent.interface';

export default function SelectionComponent(props: ISelectionComponentProps) {
	const {
		id,
		label,
		placeholder,
		values,
		onChange,
		defaultValue,
		keyValue,
		itemDisplayKey,
		overrideSelectStyles,
		overrideInputStyles,
	} = props;
	const itemKey: string = itemDisplayKey || 'name';
	const [value, setValue] = useState<string>(defaultValue || '');

	const renderValue = (selected: string) => {
		if (!selected && placeholder) {
			return <span style={{ color: 'gray' }}>{placeholder}</span>;
		}
		return selected;
	};

	const onSelection = (event: any) => {
		if (event && event.target) {
			const currentValue = (event.target as HTMLInputElement).value;

			if (onChange && keyValue) {
				onChange({
					[keyValue]: currentValue,
				});
			} else if (onChange) {
				onChange(currentValue);
			}
			setValue(currentValue);
		}
	};

	return (
		<FormControl fullWidth>
			{label ? <FormLabel component='legend'>{label}</FormLabel> : null}
			<Select
				id={id || 'selection-component'}
				input={<OutlinedInput sx={overrideInputStyles || {}} />}
				renderValue={renderValue}
				displayEmpty
				value={value}
				sx={overrideSelectStyles || {}}
				data-testid='selection-test-component'
				onChange={onSelection}
			>
				{values && values.length
					? values.map((item: any, index: number) => (
							<MenuItem key={`selection-value-${index}`} value={item.id}>
								{item[itemKey]}
							</MenuItem>
					  ))
					: []}
			</Select>
		</FormControl>
	);
}
