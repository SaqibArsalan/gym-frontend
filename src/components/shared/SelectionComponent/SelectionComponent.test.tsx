import React from 'react';
import "@testing-library/react/dont-cleanup-after-each";
import {cleanup, render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom'; 
import SelectionComponent from './SelectionComponent';

let renderWrapper: RenderResult;


describe("SelectionComponent tests", () => {
    beforeAll(() => {
        renderWrapper = render(<SelectionComponent values={[
            {id: 1, name: 'test1'},
            {id: 2, name: 'test2'}
        ]} />);
    });
    afterAll(() => {
        cleanup();
    });
    test("default renders successfully", () => {
        expect(renderWrapper.container).not.toBeNull();
        expect(renderWrapper.container).toMatchSnapshot();
    });
})