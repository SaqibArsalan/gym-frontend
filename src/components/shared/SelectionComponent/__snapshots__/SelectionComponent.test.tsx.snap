// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectionComponent tests default renders successfully 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
  >
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      data-testid="selection-test-component"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="selection-component"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="selection-component"
        role="button"
        tabindex="0"
      >
        <span
          class="notranslate"
        >
          ​
        </span>
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value=""
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-hdw1oc"
        >
          <span
            class="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
